pkg/stanza/fileconsumer/header.go

```diff
@@ -0,0 +1,274 @@
+// have its HeaderAttributes field set to the resultant header attributes.
+func (h *header) ReadHeader(ctx context.Context, f io.ReadSeeker, maxLineSize int, enc helper.Encoding, fileAttributes *FileAttributes) {
+	if h.finalized {
+		return
+	}
+
+	// Seek to the end of the last read header line
+	if _, err := f.Seek(h.offset, io.SeekStart); err != nil {
+		h.logger.Errorw("Failed to seek", zap.Error(err))
+		return
+	}
+
+	scanner := NewPositionalScanner(f, maxLineSize, h.offset, h.config.splitFunc)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		ok := scanner.Scan()
+		if !ok {
+			if err := scanner.getError(); err != nil {
+				h.logger.Errorw("Failed during header scan", zap.Error(err))
+				h.offset = scanner.Pos()
+			}
+			break
+		}
+
+		b := scanner.Bytes()
+		line, err := enc.Decode(b)
+		if err != nil {
+			h.logger.Errorw("Failed to decode header bytes", zap.Error(err))
+			h.offset = scanner.Pos()
+			continue
+		}
+
+		// If the regex doesn't match, we've reached the end of the header, so we'll finalize.
+		if !h.config.matchRegex.Match(line) {
+			h.finalizeHeader(fileAttributes)
+			return
+		}
+
+		if err := h.processHeaderLine(ctx, line); err != nil {
+			h.logger.Errorw("Failed to process header line", zap.Error(err))
+		}
+		h.offset = scanner.Pos()
+	}
+}
+
+// finalizeHeader marks the header as completely read, and adds the resultant attributes to the FileAttributes.
+func (h *header) finalizeHeader(fileAttributes *FileAttributes) {
+	h.finalized = true
+	fileAttributes.HeaderAttributes = h.attributesFromHeader
+}
+
+// processHeaderLine processes a header line, upserting entries from the line into the header attributes
+func (h *header) processHeaderLine(ctx context.Context, line []byte) error {
+	firstOperator := h.headerPipeline.Operators()[0]
+
+	newEntry := entry.New()
+	newEntry.Body = string(line)
+
+	if err := firstOperator.Process(ctx, newEntry); err != nil {
+		return fmt.Errorf("failed to process header entry: %w", err)
+	}
+
+	ent, err := h.outputOperator.WaitForEntry(ctx)
+	if err != nil {
+		return fmt.Errorf("got error while waiting for header entry: %w", err)
+	}
+
+	// Copy resultant attributes over current set of attributes (upsert)
+	for k, v := range ent.Attributes {
+		h.attributesFromHeader[k] = v
+	}
+
+	return nil
+}
+
+func (h *header) Shutdown() error {
+	return h.headerPipeline.Stop()
+}
+
+// Finalized returns true if the header is "finalized", meaning the header has been completely
+// read and processed.
+func (h header) Finalized() bool {
+	return h.finalized
+}
+
+// Offset returns the current offset into the file that the header has been read.
+// if Finalized is true, this offset indicates where the header ends for the current file.
+func (h header) Offset() int64 {
+	return h.offset
+}
+
+// headerPipelineOutput is a stanza operator that emits log entries to a channel
+type headerPipelineOutput struct {
```

Comment:
Can we reuse [`adapter.LogEmitter`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/88a8d74de64e6b7d0bed727364527ad7ecf1b100/pkg/stanza/adapter/emitter.go#L30)?

Reply:
I had looked at it originally, it seemed heavy handed for this purpose because we don't need the batching functionality, which adds overhead of an additional goroutine per header, as well as a ticker per-header. 

I'll remove it if you think the extra code outweighs the overhead, but that was my thought process.

Reply:
Good point on the batching. Not worth it for that reason.

The thing about this implementation that concerns me is the note about deadlocking and how we're handling it. Are you sure there is a risk here that we can't manage in a better way? I am surprised that we would need to drop a line in any case.

Reply:
I'm being a little over-defensive in the code, I think. The only reason why we'd ever hit that case is if we don't retrieve an entry on a cycle, which could happen if:
1. firstOperator.Process() returns an error, but the entry continues down the pipeline anyways (I don't THINK this is something in any operator, but I don't know that for sure)
2. WaitForEntry gets context cancelled. As the code is currently written, a cancellation only occurs on shutdown.

The adapter doesn't really have this problem because it doesn't have this assumption that the number of logs passed in being the number of logs passed out.

I think case 1 is basically impossible to detect, and there's not a good way that doesn't involve potentially losing entries for at least some log lines.

Case 2 we could maybe detect and track, but again, I think it's impossible to reach.

More than anything, the code there is just in case something changes with how that context is cancelled without completely locking up the filereceiver.

Reply:
I added a comment to explain the main concern here.